File: frontend/types/stripe.ts
/**
 * TypeScript types for Stripe integration
 */

export interface StripeCustomer {
  id: number;
  stripe_customer_id: string;
  email: string;
  name?: string;
  created_at: string;
}

export interface StripeSubscription {
  id: number;
  stripe_subscription_id: string;
  status: string;
  current_period_start: string;
  current_period_end: string;
  amount: number;
  currency: string;
  product_name: string;
  interval: string;
  interval_count: number;
}

export interface StripePaymentIntent {
  id: number;
  stripe_payment_intent_id: string;
  amount: number;
  currency: string;
  status: string;
  description?: string;
  created_at: string;
}

export interface StripeProduct {
  id: string;
  name: string;
  description?: string;
  active: boolean;
}

export interface StripePrice {
  id: string;
  unit_amount: number;
  currency: string;
  recurring?: {
    interval: string;
    interval_count: number;
  };
}

export interface StripeProductWithPrices {
  id: string;
  name: string;
  description?: string;
  prices: StripePrice[];
}

export interface CheckoutSessionCreate {
  user_id: number;
  price_id: string;
  success_url: string;
  cancel_url: string;
  mode: 'payment' | 'subscription';
  metadata?: Record<string, string>;
}

export interface CheckoutSessionResponse {
  session_id: string;
  url: string;
}

export interface SubscriptionCreate {
  user_id: number;
  price_id: string;
  payment_method_id?: string;
  trial_period_days?: number;
}

export interface SubscriptionResponse {
  id: number;
  stripe_subscription_id: string;
  status: string;
  current_period_start: string;
  current_period_end: string;
  amount: number;
  currency: string;
  product_name: string;
}

export interface CustomerPortalResponse {
  url: string;
}

export interface StripeWebhookEvent {
  id: number;
  stripe_event_id: string;
  event_type: string;
  processed: boolean;
  created_at: string;
  processed_at?: string;
  error_message?: string;
}

// Payment form types
export interface PaymentFormProps {
  amount: number;
  currency: string;
  onSuccess: (paymentResult: any) => void;
  onError: (error: string) => void;
  description?: string;
  metadata?: Record<string, string>;
}

// Subscription manager types
export interface SubscriptionManagerProps {
  userId: number;
}

// Financial dashboard types
export interface FinancialOverview {
  total_users: number;
  total_credits: number;
  active_subscriptions: number;
  total_subscription_revenue: number;
}

export interface MonthlyStats {
  credits_spent: number;
  credits_added: number;
  net_change: number;
}

export interface Transaction {
  id: string;
  user_email: string;
  amount: number;
  description: string;
  created_at: string;
  type: 'credit' | 'debit';
}

export interface FinancialDashboardData {
  overview: FinancialOverview;
  monthly_stats: MonthlyStats;
  recent_transactions: Transaction[];
}

// Credit management types
export interface CreditBalance {
  current_balance: number;
  total_earned: number;
  total_spent: number;
  recent_transactions: Transaction[];
}

export interface CreditTransaction {
  id: string;
  amount: number;
  description: string;
  created_at: string;
  type: 'credit' | 'debit';
}

// API response types
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  message?: string;
  error?: string;
}

export interface StripeError {
  code: string;
  message: string;
  details?: any;
}

File: backend/schemas/stripe.py
"""
Pydantic schemas for Stripe API requests and responses.
"""
from pydantic import BaseModel, HttpUrl
from typing import Optional, Dict, Any
from datetime import datetime


class CheckoutSessionCreate(BaseModel):
    """Request schema for creating a checkout session"""
    user_id: int
    price_id: str
    success_url: HttpUrl
    cancel_url: HttpUrl
    mode: str = "subscription"  # "payment" or "subscription"
    metadata: Optional[Dict[str, Any]] = None


class CheckoutSessionResponse(BaseModel):
    """Response schema for checkout session creation"""
    session_id: str
    url: HttpUrl


class SubscriptionCreate(BaseModel):
    """Request schema for creating a subscription"""
    user_id: int
    price_id: str
    payment_method_id: Optional[str] = None
    trial_period_days: Optional[int] = None


class SubscriptionResponse(BaseModel):
    """Response schema for subscription creation"""
    id: int
    stripe_subscription_id: str
    status: str
    current_period_start: datetime
    current_period_end: datetime
    amount: float
    currency: str
    product_name: str


class CustomerPortalResponse(BaseModel):
    """Response schema for customer portal URL"""
    url: HttpUrl


class StripeProduct(BaseModel):
    """Schema for Stripe product information"""
    id: str
    name: str
    description: Optional[str] = None
    active: bool = True


class StripePrice(BaseModel):
    """Schema for Stripe price information"""
    id: str
    unit_amount: int
    currency: str
    recurring: Optional[Dict[str, Any]] = None


class StripeProductWithPrices(BaseModel):
    """Schema for product with associated prices"""
    id: str
    name: str
    description: Optional[str] = None
    prices: list[StripePrice]


class StripeSubscription(BaseModel):
    """Schema for subscription information"""
    id: int
    stripe_subscription_id: str
    status: str
    current_period_start: datetime
    current_period_end: datetime
    amount: float
    currency: str
    product_name: str
    interval: str
    interval_count: int


class StripeCustomer(BaseModel):
    """Schema for customer information"""
    id: int
    stripe_customer_id: str
    email: str
    name: Optional[str] = None
    created_at: datetime


class StripePaymentMethod(BaseModel):
    """Schema for payment method information"""
    id: int
    stripe_payment_method_id: str
    type: str
    card_brand: Optional[str] = None
    card_last4: Optional[str] = None
    card_exp_month: Optional[int] = None
    card_exp_year: Optional[int] = None
    is_default: bool = False
    is_active: bool = True


class StripeWebhookEvent(BaseModel):
    """Schema for webhook event information"""
    id: int
    stripe_event_id: str
    event_type: str
    processed: bool
    created_at: datetime
    processed_at: Optional[datetime] = None
    error_message: Optional[str] = None
